README.MD
# ISO8601

[![Build Status](https://travis-ci.org/sagotch/ISO8601.ml.svg?branch=test)](https://travis-ci.org/sagotch/ISO8601.ml)

ISO 8601 and RFC 3999 date parsing for OCaml.

**Work in progress. API should not be considerer stable
until 1.0.0 version.**

## Permissive module

Date, time, and datetime parsing function of this module are
much more permissive than ISO 8601 or RFC 3339 specifications.

Either basic and/or extended formats can be used for date and time,
and no consistency is required (it is possible to mix basic format for
date and extended format for time), lowercase letters can be used,
and datetime separator may be a space character.

Note: for now, this module is the only one available.

## Installation

Via opam:

    opam install ISO8601

From sources:

    make build
    make install

## License

Distributed under the terms of [MIT license](LISENCE).


CHANGELOG.MD
# CHANGELOG

## 0.2.5

- Added unix requirement in META file

## 0.2.4

- "pp_format: the 'minute' specifier is now modulo 60"
  (patch by Erkki Seppälä)

## 0.2.3

- Fixed zero padding when printing time zone offset.

## 0.2.2

- Fixed "Warning 40: tm_isdst was selected from type Unix.tm"

## 0.2.1

- Fixing the previous important bugfix. (!)
- Fixing the daylight time saving offset in timestamp.

## 0.2.0

- Breaking change in API: functions with optionnal time zone
  argument splitted into 2 different functions. With and without
  time zone argument.
- Used format-like function to be able to customize printing.
- Added printers for basic format.
- Important bugfix when using mktime.

## 0.1.3

- Printer bugfix.

## 0.1.2

- Printer bugfix.

## 0.1.1

- Fixed OCaml 3.12.1 compatibility.

## 0.1.0

- Permissive module will parse a superset of (a part of)
  the ISO8601 spec.
- Printing of date / times / ... with the form defined as
  the extended format.


test/test.ml
(*
 * Copyright (c) 2015 David Sheets <sheets@alum.mit.edu>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 *)

module Tm_struct : Alcotest.TESTABLE with type t = Unix.tm = struct
  type t = Unix.tm

  let pp fmt tm =
    let open Unix in
    let s = Printf.sprintf "%d-%02d-%02dT%02d:%02d:%02d"
        (1900+tm.tm_year) (tm.tm_mon+1) tm.tm_mday
        tm.tm_hour tm.tm_min tm.tm_sec
    in
    Format.pp_print_string fmt s

  let equal a b =
    Unix.(a.tm_sec = b.tm_sec
          && a.tm_min = b.tm_min
          && a.tm_hour = b.tm_hour
          && a.tm_mday = b.tm_mday
          && a.tm_mon = b.tm_mon
          && a.tm_year = b.tm_year)
end

let tm_struct = (module Tm_struct : Alcotest.TESTABLE with type t = Unix.tm)

type hemi = Neg | Pos
type tz = Local | Z | Tz of hemi * int * int

let est = Tz (Neg, 5, 0)
let ist = Tz (Pos, 5, 30)
let vet = Tz (Neg, 4, 30)

let time_tests f = [
  "before_1900",      `Quick, f 1861  9  1 8 0 0 Local;
  "before_epoch",     `Quick, f 1969  1  1 0 0 1 Local;
  "nowish",           `Quick, f 2015 12 27 0 0 1 Local;
  "before_1900_z",    `Quick, f 1861  9  1 8 0 0 Z;
  "before_epoch_z",   `Quick, f 1969  1  1 0 0 1 Z;
  "nowish_z",         `Quick, f 2015 12 27 0 0 1 Z;
  "before_1900_est",  `Quick, f 1861  9  1 8 0 0 est;
  "before_epoch_est", `Quick, f 1969  1  1 0 0 1 est;
  "nowish_est",       `Quick, f 2015 12 27 0 0 1 est;
  "before_1900_ist",  `Quick, f 1861  9  1 8 0 0 ist;
  "before_epoch_ist", `Quick, f 1969  1  1 0 0 1 ist;
  "nowish_ist",       `Quick, f 2015 12 27 0 0 1 ist;
  "before_1900_vet",  `Quick, f 1861  9  1 8 0 0 vet;
  "before_epoch_vet", `Quick, f 1969  1  1 0 0 1 vet;
  "nowish_vet",       `Quick, f 2015 12 27 0 0 1 vet;
]

let fixed_time_tests f = [
  "fixed_unix_time_nowish_utc", `Quick,
  f 1451407335. 0.        "2015-12-29T16:42:15Z";
  "fixed_unix_time_nowish_est", `Quick,
  f 1451407335. (-18000.) "2015-12-29T11:42:15-05:00";
  "fixed_unix_time_nowish_ist", `Quick,
  f 1451407335. 19800.    "2015-12-29T22:12:15+05:30";
  "fixed_unix_time_nowish_vet", `Quick,
  f 1451407335. (-16200.) "2015-12-29T12:12:15-04:30";
]

let str_tm year month day hour minute second tz =
  let str = Printf.sprintf "%d-%02d-%02dT%02d:%02d:%02d%s"
      year month day hour minute second
      (match tz with
       | Local -> ""
       | Z -> "Z"
       | Tz (Neg, hr, mn) -> Printf.sprintf "-%02d:%02d" hr mn
       | Tz (Pos, hr, mn) -> Printf.sprintf "+%02d:%02d" hr mn
      )
  in
  let tm = Unix.({
    tm_sec  = second;
    tm_min  = minute;
    tm_hour = hour;
    tm_mday = day;
    tm_mon  = month - 1;
    tm_year = year - 1900;
    tm_wday = 0;
    tm_yday = 0;
    tm_isdst = false;
  }) in
  let t,  _ = Unix.mktime tm in
  let t', _ = Unix.mktime (Unix.gmtime t) in
  let local_unix_time = t -. (t' -. t) in
  let tm = match tz with
    | Local -> tm
    | Z -> snd Unix.(mktime (gmtime local_unix_time))
    | Tz (Neg, hr, mn) ->
      let t = local_unix_time +. (float_of_int ((hr * 3600) + (mn * 60))) in
      snd Unix.(mktime (gmtime t))
    | Tz (Pos, hr, mn) ->
      let t = local_unix_time -. (float_of_int ((hr * 3600) + (mn * 60))) in
      snd Unix.(mktime (gmtime t))
  in
  (str, tm)

let erange = Unix.Unix_error (Unix.ERANGE, "mktime", "")

let parse_test year month day hour minute second tz () =
  if year < 1900
  then Alcotest.check_raises "< 1900 is ERANGE" erange (fun () ->
    ignore (ISO8601.Permissive.datetime "1861-01-01T00:00:00Z")
  )
  else
    let str, tm = str_tm year month day hour minute second tz in
    let parsed = ISO8601.Permissive.datetime str in
    let output = match tz with
      | Local    -> Unix.localtime parsed
      | Z | Tz _ -> Unix.gmtime parsed
    in
    Alcotest.(check tm_struct ("parse "^str) tm output)

let parse_fixed_unix_time unix_time tz s () =
  let parsed = int_of_float (ISO8601.Permissive.datetime s) in
  Alcotest.(check int ("parse "^s) (int_of_float unix_time) parsed)

let parse_tests =
  time_tests parse_test @ fixed_time_tests parse_fixed_unix_time

let string_of_datetime unix_time = function
  | Local -> ISO8601.Permissive.string_of_datetime unix_time
  | Z -> ISO8601.Permissive.string_of_datetimezone (unix_time,0.)
  | Tz (Neg,hr,mn) ->
    let tz = float_of_int (- (hr * 3600 + mn * 60)) in
    ISO8601.Permissive.string_of_datetimezone (unix_time, tz)
  | Tz (Pos,hr,mn) ->
    let tz = float_of_int (hr * 3600 + mn * 60) in
    ISO8601.Permissive.string_of_datetimezone (unix_time, tz)

let print_test year month day hour minute second tz () =
  (* We use Unix.mktime to find the epoch time but with year < 1900 it
     will error with ERANGE. *)
  if year < 1900
  then ()
  else
    let str, tm = str_tm year month day hour minute second tz in
    let unix_time, _ = Unix.mktime tm in
    let unix_time = match tz with
      | Local -> unix_time
      | Z | Tz _ ->
        let offt, _ = Unix.mktime (Unix.gmtime unix_time) in
        unix_time -. (offt -. unix_time)
    in
    let output = string_of_datetime unix_time tz in
    Alcotest.(check string ("print "^str) str output)

let print_fixed_unix_time unix_time tz s () =
  let output = ISO8601.Permissive.string_of_datetimezone (unix_time, tz) in
  Alcotest.(check string ("print "^s) s output)

let print_tests =
  time_tests print_test @ fixed_time_tests print_fixed_unix_time

let rt_test year month day hour minute second tz () =
  if year < 1900
  then Alcotest.check_raises "< 1900 is ERANGE" erange (fun () ->
    ignore (ISO8601.Permissive.datetime "1861-01-01T00:00:00")
  )
  else
    let str, _ = str_tm year month day hour minute second tz in
    let output = string_of_datetime (ISO8601.Permissive.datetime str) tz in
    Alcotest.(check string ("roundtrip "^str) str output)

let rt_fixed_unix_time unix_time tz s () =
  let output = ISO8601.Permissive.(string_of_datetimezone (datetime s, tz)) in
  Alcotest.(check string ("roundtrip "^s) s output);
  let output = int_of_float ISO8601.Permissive.(
    datetime (string_of_datetimezone (unix_time, tz))
  ) in
  let unix_time = int_of_float unix_time in
  Alcotest.(check int ("roundtrip "^string_of_int unix_time) unix_time output)

let rt_tests =
  time_tests rt_test @ fixed_time_tests rt_fixed_unix_time

let suites = [
  "parse", parse_tests;
  "print", print_tests;
  "rt",    rt_tests;
]

;;
Alcotest.run "ISO8601" suites


UTILS_TEST.ml
let test a b =
  OUnit.(>::)
        (string_of_float a)
        (fun _ -> OUnit.assert_equal
                    ~cmp:(OUnit.cmp_float ~epsilon:Pervasives.epsilon_float)
                    ~printer:string_of_float
                    a b)

let _ =
  let mkdatetime = Utils.mkdatetime in
  let mkdate = Utils.mkdate in
  let mktime = Utils.mktime in
  [
    OUnit.(>:::) "[UTILS mkdatetime]"
          [
            test 0. (mkdatetime 1970 1 1 0 0 0) ;
            test 1. (mkdatetime 1970 1 1 0 0 1) ;
            test 60. (mkdatetime 1970 1 1 0 1 0) ;
            test 3600. (mkdatetime 1970 1 1 1 0 0) ;
            test (31. *. 24. *. 3600.)
                         (mkdatetime 1970 2 1 0 0 0) ;
            test (365. *. 24. *. 3600.) (* not leap year *)
                         (mkdatetime 1971 1 1 0 0 0) ;
          ] ;
    OUnit.(>:::) "[UTILS mkdate]"
          [
            test 0. (mkdate 1970 1 1) ;
            test (24. *. 3600.) (mkdate 1970 1 2) ;
            test (31. *. 24. *. 3600.) (mkdate 1970 2 1) ;
            test 583804800. (mkdate 1988 7 2) ;
          ] ;
    OUnit.(>:::) "[UTILS mktime]"
          [
            test 0. (mktime 0. 0. 0.) ;
            test 60. (mktime 0. 1. 0.) ;
            test 1. (mktime 0. 0. 1.) ;
          ]
  ]
  |> List.map OUnit.run_test_tt_main


src/ISO8601.ml
module Lexer = ISO8601_lexer

module Permissive = struct

    let date_lex lexbuf = Lexer.date lexbuf

    let time_tz_lex lexbuf =
      let t = Lexer.time lexbuf in
      let tz = Lexer.timezone lexbuf in
      let t = match tz with None -> t | Some o -> t -. o in
      (t, tz)

    let datetime_tz_lex ?(reqtime=true) lexbuf =
      let d = date_lex lexbuf in
      match Lexer.delim lexbuf with
      | None ->
        (* TODO: this should be a real exception *)
        if reqtime then assert false else (d, None)
      | Some _ ->
        let (t, tz) = time_tz_lex lexbuf in
        match tz with
        | None -> (d +. t, tz)
        | Some tz ->
          let t = d +. t in
          let offt = fst (Unix.mktime (Unix.gmtime t)) in
          (t -. (offt -. t), Some tz)

    let time_lex lexbuf =
      fst (time_tz_lex lexbuf)

    let datetime_lex ?(reqtime=true) lexbuf =
      fst (datetime_tz_lex ~reqtime:reqtime lexbuf)

    let date s = date_lex (Lexing.from_string s)

    let time s = time_lex (Lexing.from_string s)

    let time_tz s = time_tz_lex (Lexing.from_string s)

    let datetime_tz ?(reqtime=true) s =
      datetime_tz_lex ~reqtime:reqtime (Lexing.from_string s)

    let datetime ?(reqtime=true) s =
      datetime_lex ~reqtime:reqtime (Lexing.from_string s)

    let pp_format fmt format x tz =

      let open Unix in
      let open Format in

      (* Be careful, do not forget to print timezone if there is one,
       * or information printed will be wrong. *)
      let x = match tz with
        | None    -> localtime x
        | Some tz -> gmtime (x +. tz)
      in

      let print_tz_hours fmt tz =
        fprintf fmt "%0+3d" (Pervasives.truncate (tz /. 3600.))
      in

      let print_tz_minutes fmt tz =
        fprintf fmt "%02.0f" (mod_float (abs_float (tz /. 60.)) 60.0)
      in

      let conversion =
        let pad2 = fprintf fmt "%02d" in
        let pad4 = fprintf fmt "%04d" in
        function

        (* Date *)
        | 'Y' -> pad4 (x.tm_year + 1900)
        | 'M' -> pad2 (x.tm_mon + 1)
        | 'D' -> pad2 x.tm_mday

        (* Time *)
        | 'h' -> pad2 x.tm_hour
        | 'm' -> pad2 x.tm_min
        | 's' -> pad2 x.tm_sec

        (* Timezone *)
        | 'Z' -> begin match tz with (* with colon *)
          | None    -> ()
          | Some 0. -> fprintf fmt "Z"
          | Some tz ->
            print_tz_hours fmt tz;
            fprintf fmt ":";
            print_tz_minutes fmt tz
        end
        | 'z' -> begin match tz with (* without colon *)
          | None    -> ()
          | Some 0. -> fprintf fmt "Z"
          | Some tz ->
            print_tz_hours fmt tz;
            print_tz_minutes fmt tz
        end

        | '%' -> pp_print_char fmt '%'
        |  c  -> failwith ("Bad format: %" ^ String.make 1 c)

      in

      let len = String.length format in
      let rec parse_format i =
        if i = len then ()
        else match String.get format i with
             | '%' -> conversion (String.get format (i + 1)) ;
                      parse_format (i + 2)
             |  c  -> pp_print_char fmt c ;
                      parse_format (i + 1) in

      parse_format 0

    let pp_date_utc fmt x = pp_format fmt "%Y-%M-%D" x (Some 0.)
    let pp_date     fmt x = pp_format fmt "%Y-%M-%D" x None

    let pp_time_utc fmt x = pp_format fmt "%h:%m:%s" x (Some 0.)
    let pp_time     fmt x = pp_format fmt "%h:%m:%s" x None

    let pp_datetime_utc fmt x = pp_format fmt "%Y-%M-%DT%h:%m:%s" x (Some 0.)
    let pp_datetime     fmt x = pp_format fmt "%Y-%M-%DT%h:%m:%s" x None

    let pp_datetimezone fmt (x, tz) =
      pp_format fmt "%Y-%M-%DT%h:%m:%s%Z" x (Some tz)

    let pp_date_basic_utc fmt x = pp_format fmt "%Y%M%D" x (Some 0.)
    let pp_date_basic     fmt x = pp_format fmt "%Y%M%D" x None

    let pp_time_basic_utc fmt x = pp_format fmt "%h%m%s" x (Some 0.)
    let pp_time_basic     fmt x = pp_format fmt "%h%m%s" x None

    let pp_datetime_basic_utc fmt x = pp_format fmt "%Y%M%DT%h%m%s" x (Some 0.)
    let pp_datetime_basic     fmt x = pp_format fmt "%Y%M%DT%h%m%s" x None

    let pp_datetimezone_basic fmt (x, tz) =
      pp_format fmt "%Y%M%DT%h%m%s%z" x (Some tz)

    let string_of_aux printer x =
      ignore (Format.flush_str_formatter ()) ;
      printer Format.str_formatter x ;
      Format.flush_str_formatter ()

    let string_of_date_utc = string_of_aux pp_date_utc
    let string_of_date     = string_of_aux pp_date

    let string_of_time_utc = string_of_aux pp_time_utc
    let string_of_time     = string_of_aux pp_time

    let string_of_datetime_utc = string_of_aux pp_datetime_utc
    let string_of_datetime     = string_of_aux pp_datetime

    let string_of_datetimezone = string_of_aux pp_datetimezone

    let string_of_date_basic_utc = string_of_aux pp_date_basic_utc
    let string_of_date_basic     = string_of_aux pp_date_basic

    let string_of_time_basic_utc = string_of_aux pp_time_basic_utc
    let string_of_time_basic     = string_of_aux pp_time_basic

    let string_of_datetime_basic_utc = string_of_aux pp_datetime_basic_utc
    let string_of_datetime_basic     = string_of_aux pp_datetime_basic

    let string_of_datetimezone_basic = string_of_aux pp_datetimezone_basic

end


ISO8601_PARSER.ml
let test p fn input expected =
  let result = fn input in
  let assert_equal = OUnit2.assert_equal
                       ~cmp:(OUnit2.cmp_float ~epsilon:Pervasives.epsilon_float)
                       ~printer:p in
  OUnit2.(>::) input (fun _ -> assert_equal expected result)

let date = test ISO8601.Permissive.string_of_date ISO8601.Permissive.date

let time = test ISO8601.Permissive.string_of_time ISO8601.Permissive.time

let datetime = test ISO8601.Permissive.string_of_datetime ISO8601.Permissive.datetime

(* Parser tests *)
let _ =
  let mkdatetime = Utils.mkdatetime in
  let mkdate = Utils.mkdate in
  let mktime = Utils.mktime in
  [
    OUnit2.(>:::) "[DATE]"
          [
            date "1970-01-01" 0. ;
            date "19700101" 0. ;
            date "1970-01" 0. ;
            date "197001" 0. ;
            date "1970" 0. ;
            date "2009" (mkdate 2009 1 1) ;
            date "2009-05-19" (mkdate 2009 5 19) ;
            date "20090519" (mkdate 2009 5 19) ;
            date "2009-05" (mkdate 2009 5 1) ;
            date "2009-001" (mkdate 2009 1 1);
            date "2009-123" (mkdate 2009 5 3);
            date "2009-222" (mkdate 2009 8 10);
            date "2009-139" (mkdate 2009 5 19);
            date "2012-060" (mkdate 2012 2 29); (* leap year *)
          ] ;
    OUnit2.(>:::) "[TIME WITHOUT TIMEZONE]"
          [
            time "12:34" (mktime 12. 34. 0.) ;
            time "00:00" (mktime 0. 0. 0.) ;
            time "14" (mktime 14. 0. 0.) ;
            time "14:31" (mktime 14. 31. 0.) ;
            time "14:39:22" (mktime 14. 39. 22.) ;
            time "24:00" (mktime 24. 0. 0.) ;
            time "16:23:48.5" (mktime 16. 23. 48.5) ;
            time "16:23:48,444" (mktime 16. 23. 48.444) ;
            time "16:23.4" (mktime 16. 23.4 0.) ;
            time "16:23,25" (mktime 16. 23.25 0.);
            time "16.23334444" (mktime 16.23334444 0. 0.);
            time "16,2283" (mktime 16.2283 0. 0.);
          ] ;
    OUnit2.(>:::) "[TIME WITH TIMEZONE]"
          [
            time "14:39:22-06:00" (mktime 20. 39. 22.) ;
            time "14:39:22+0600" (mktime 8. 39. 22.);
            time "14:39:22-01" (mktime 15. 39. 22.);
            time "0545Z" (mktime 5. 45. 0.);
            time "16:23:48,3-06:00" (mktime 22. 23. 48.3) ;
            time "16:23.33+0600" (mktime 10. 23.33 0.) ;
          ] ;
    OUnit2.(>:::) "[DATETIME WITHOUT TIMEZONE]"
          [
            datetime "2015-02-15T11:55" (mkdatetime 2015 02 15 11 55 0) ;
          ] ;
  ]
  |> List.map OUnit2.run_test_tt_main


ISO8601_PRINTER.ml
let test (fn : float -> string) (input : float) (expected : string)  =
  let result = fn input in
  let assert_equal = OUnit.assert_equal ~printer:(fun x -> x) in
  OUnit.(>::) (string_of_float input)
        (fun _ -> assert_equal expected result)

let date = test ISO8601.Permissive.string_of_date

let time = test ISO8601.Permissive.string_of_time

let datetime = test ISO8601.Permissive.string_of_datetime

let _ =
  let mkdate = Utils.mkdate in
  [
    OUnit.(>:::) "[PRINTER DATE]"
          [
            date 0. "1970-01-01" ;
            date (24. *. 3600.) "1970-01-02" ;
            date (31. *. 24. *. 3600.) "1970-02-01" ;
            date (365. *. 24. *. 3600.) "1971-01-01" ;
            date (mkdate 2009 1 1) "2009-01-01" ;
            date (mkdate 2009 5 19) "2009-05-19" ;
            date 583804800. "1988-07-02" ;
            date (mkdate 1988 7 2) "1988-07-02" ;
          ];
    OUnit.(>:::) "[PRINTER TIME]"
          [
            time 0. "00:00:00" ;
            time 1. "00:00:01" ;
            time 60. "00:01:00" ;
            time 3600. "01:00:00" ;
          ] ;
  ]
  |> List.map OUnit.run_test_tt_main
